packages/dd-trace/src/collector/span.js

```diff
@@ -0,0 +1,118 @@
+'use strict'
+
+const { channel } = require('dc-polyfill')
+const DatadogSpan = require('../opentracing/span')
+const DatadogCollectorSpanContext = require('./span_context')
+
+const now = performance.now.bind(performance)
+
+const startSegmentChannel = channel('datadog:tracing:segment:start')
+const segmentDiscardChannel = channel('datadog:tracing:segment:discard')
+const startChannel = channel('datadog:tracing:span:start')
+const tagsChannel = channel('datadog:tracing:span:tags')
+const errorChannel = channel('datadog:tracing:span:error')
+const finishChannel = channel('datadog:tracing:span:finish')
+
+let segmentId = 0
+
+class DatadogCollectorSpan extends DatadogSpan {
+  finish (finishTime) {
+    if (this._spanContext._isFinished) return
+
+    this._spanContext._isFinished = true
+    this._spanContext._trace.active--
+
+    const trace = this._spanContext._trace
+
+    // TODO: Emit a discard event from tracer. For now we just short-circuit.
+    if (trace.isDiscarded) return
+    if (trace.isRecording === false) {
+      trace.isDiscarded = true
+      return segmentDiscardChannel.publish({ segmentId })
+    }
+
+    const ticks = finishTime
+      ? now() + finishTime - trace.ticks - trace.startTime
+      : now() - trace.ticks
+
+    finishChannel.publish({
+      ticks,
+      segmentId: trace.segmentId,
+      spanIndex: this._spanContext._spanIndex
+    })
+  }
+
+  _start (_tracer, _processor, _prioritySampler, fields, _debug) {
+    const parent = fields.parent || null
+    const tags = fields.tags || {}
+    const spanContext = this._spanContext
+    const trace = spanContext._trace
+
+    this._trackSegment()
+
+    spanContext._spanIndex = trace.lastIndex++
+
+    if (parent && trace.segmentId === parent.segmentId) {
+      spanContext._parentIndex = parent._spanIndex >= 0
+        ? parent._spanIndex + 1
+        : 0
+    }
+
+    startChannel.publish({
+      ticks: now() - spanContext._trace.ticks,
+      segmentId: spanContext._trace.segmentId,
+      spanId: spanContext._spanId,
+      parentIndex: spanContext._parentIndex,
+      type: tags['span.type'],
+      name: fields.operationName,
+      resource: tags['resource.name'],
+      service: tags['service.name'],
+      meta: tags,
+      metrics: tags
+    })
+  }
+
+  _addTags (keyValuePairs = {}) {
+    tagsChannel.publish({
+      segmentId: this._spanContext._trace.segmentId,
+      spanIndex: this._spanContext._spanIndex,
+      meta: keyValuePairs,
+      metrics: keyValuePairs
+    })
+
+    if (keyValuePairs.error) {
+      errorChannel.publish({
+        segmentId: this._spanContext._trace.segmentId,
+        spanIndex: this._spanContext._spanIndex,
+        error: keyValuePairs.error
+      })
+    }
+  }
+
+  _trackSegment () {
```

Comment:
Why do spans need awareness of segments now? Can we avoid this somehow and leave that up to the collector?

Reply:
Spans have always had awareness of segments as the perhaps not greatly named `_trace` property. It _must_ know about it at the very minimum for correlation purpose.

Reply:
We need some IDs for correlation, but we shouldn't need to hold the whole trace. Seems to me like we're not getting any of the benefits of event-based output if we're still holding all the same data for the life of the spans or traces. 🤔

Reply:
> We need some IDs for correlation, but we shouldn't need to hold the whole trace. Seems to me like we're not getting any of the benefits of event-based output if we're still holding all the same data for the life of the spans or traces.

All it's keeping is the IDs, plus a few integers used only for state tracking. It doesn't hold any actual trace data, only what is needed to know the current status of the segment, so I fail to understand your point.

Reply:
That is a temporary implementation detail while both approaches have to continue working in parallel. A lot of the code relies on random things from the SpanContext today. The idea of the PR is to at least enable reporting the data as events, not to solve all the underlying problem with the rest of the code that prevent moving the state out entirely. For the events themselves however, only the IDs and active span count are needed. Removing most of the state from the process entirely will likely be several months or even quarters of work.